asyncapi/document.go

```diff
@@ -2,17 +2,150 @@ package asyncapi
+// NOTE: this interface is not completed yet.
 type Document interface {
 	Extendable
+	Channels() []Channel
+	HasChannels() bool
+	ApplicationPublishableChannels() []Channel
+	ApplicationPublishableMessages() []Message
+	ApplicationPublishOperations() []Operation
+	ApplicationSubscribableChannels() []Channel
+	ApplicationSubscribableMessages() []Message
+	ApplicationSubscribeOperations() []Operation
+	ClientPublishableChannels() []Channel
+	ClientPublishableMessages() []Message
+	ClientPublishOperations() []Operation
+	ClientSubscribableChannels() []Channel
+	ClientSubscribableMessages() []Message
+	ClientSubscribeOperations() []Operation
+	Messages() []Message
 	Server(name string) (Server, bool)
 	Servers() []Server
 	HasServers() bool
 }
 
+// Channel is an addressable component, made available by the server, for the organization of messages.
+// Producer applications send messages to channels and consumer applications consume messages from channels.
+type Channel interface {
+	Extendable
+	Identifiable
+	Path() string // Path is the identifier
+	Description() string
+	HasDescription() bool
+	Parameters() []ChannelParameter
+	HasParameters() bool
+	Operations() []Operation
+	Messages() []Message
+}
+
+// ChannelParameter describes a parameter included in a channel name.
+type ChannelParameter interface {
+	Extendable
+	Identifiable
+	Name() string
+	Description() string
+	Schema() Schema
+}
+
+// OperationType is the type of an operation.
+type OperationType string
+
+// Operation describes a publish or a subscribe operation.
+// This provides a place to document how and why messages are sent and received.
+type Operation interface {
+	Extendable
+	ID() string
+	Description() string
+	HasDescription() bool
+	IsApplicationPublishing() bool
+	IsApplicationSubscribing() bool
+	IsClientPublishing() bool
+	IsClientSubscribing() bool
+	Messages() []Message
+	Summary() string
+	HasSummary() bool
+	Type() OperationType
+}
+
+// Message describes a message received on a given channel and operation.
+type Message interface {
+	Extendable
+	UID() string
+	Name() string
+	Title() string
+	HasTitle() bool
+	Description() string
+	HasDescription() bool
+	Summary() string
+	HasSummary() bool
+	ContentType() string
+	Payload() Schema
+}
+
+// Schema is an object that allows the definition of input and output data types.
+// These types can be objects, but also primitives and arrays.
+// This object is a superset of the JSON Schema Specification Draft 07.
+type Schema interface {
+	Extendable
+	ID() string
+	AdditionalItems() Schema
+	AdditionalProperties() Schema // TODO (boolean | Schema)
+	AllOf() []Schema
+	AnyOf() []Schema
+	CircularProps() []string
+	Const() interface{}
+	Contains() Schema
+	ContentEncoding() string
+	ContentMediaType() string
+	Default() interface{}
+	Definitions() map[string]Schema
+	Dependencies() map[string]Schema // TODO Map[string, Schema|string[]]
+	Deprecated() bool
+	Description() string
+	HasDescription() bool
```

Comment:
Schema shouldn't have the `HasDescription` function for me. If you want leave it, then you should also make (or update the parser-api) the `Has*` function for all JSON Schema keywords.

Reply:
Good catch! removed